File: CPP/DestroyingAsteroids.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <vector>

class Solution {
public:
    bool asteroidsDestroyed(int mass, std::vector<int>& asteroids) {
        std::sort(asteroids.begin(), asteroids.end());
        // Sum may exceed int32 (~1e10); use int64.
        int64_t m = mass;
        for (int a : asteroids) {
            if (m < a) return false;
            m += a;
        }
        return true;
    }
};

int main() {
    Solution s;
    std::vector<int> a = {3, 9, 19, 5, 21};
    std::vector<int> b = {4, 9, 23, 4};
    std::vector<int> c = {1};
    std::vector<int> d = {2};
    std::cout << std::boolalpha;
    std::cout << s.asteroidsDestroyed(10, a) << "\n"; // true
    std::cout << s.asteroidsDestroyed(5, b) << "\n";  // false
    std::cout << s.asteroidsDestroyed(1, c) << "\n";  // true
    std::cout << s.asteroidsDestroyed(1, d) << "\n";  // false
    return 0;
}

File: Rust/src/bin/destroying_asteroids.rs
fn asteroids_destroyed(mass: i32, mut asteroids: Vec<i32>) -> bool {
    asteroids.sort_unstable();
    // Sum of asteroids can reach 1e10, so accumulate in i64.
    let mut m = mass as i64;
    for a in asteroids {
        if m < a as i64 {
            return false;
        }
        m += a as i64;
    }
    true
}

fn main() {
    println!("{}", asteroids_destroyed(10, vec![3, 9, 19, 5, 21])); // true
    println!("{}", asteroids_destroyed(5, vec![4, 9, 23, 4]));      // false
    println!("{}", asteroids_destroyed(1, vec![1]));                // true
    println!("{}", asteroids_destroyed(1, vec![2]));                // false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert!(asteroids_destroyed(10, vec![3, 9, 19, 5, 21]));
        assert!(!asteroids_destroyed(5, vec![4, 9, 23, 4]));
        assert!(asteroids_destroyed(1, vec![1]));
        assert!(!asteroids_destroyed(1, vec![2]));
    }
}
